src/Symfony/Component/DependencyInjection/Tests/Compiler/FactoryReturnTypePassTest.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\FactoryReturnTypePass;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\DependencyInjection\Tests\Fixtures\factoryFunction;
+use Symfony\Component\DependencyInjection\Tests\Fixtures\FactoryDummy;
+use Symfony\Component\DependencyInjection\Tests\Fixtures\FactoryParent;
+
+/**
+ * @author Guilhem N. <egetick@gmail.com>
+ */
+class FactoryReturnTypePassTest extends \PHPUnit_Framework_TestCase
+{
+    public function testProcess()
+    {
+        $container = new ContainerBuilder();
+
+        $factory = $container->register('factory');
+        $factory->setFactory(array(FactoryDummy::class, 'createFactory'));
+
+        $foo = $container->register('foo');
+        $foo->setFactory(array(new Reference('factory'), 'create'));
+
+        $bar = $container->register('bar', __CLASS__);
+        $bar->setFactory(array(new Reference('factory'), 'create'));
+
+        $pass = new FactoryReturnTypePass();
+        $pass->process($container);
+
+        if (method_exists(\ReflectionMethod::class, 'getReturnType')) {
+            $this->assertEquals(FactoryDummy::class, $factory->getClass());
+            $this->assertEquals(\stdClass::class, $foo->getClass());
+        } else {
+            $this->assertNull($factory->getClass());
+            $this->assertNull($foo->getClass());
+        }
+        $this->assertEquals(__CLASS__, $bar->getClass());
+    }
+
+    /**
+     * @dataProvider returnTypesProvider
+     */
+    public function testReturnTypes($factory, $returnType, $hhvmSupport = true)
+    {
+        if (!$hhvmSupport && defined('HHVM_VERSION')) {
+            $this->markTestSkipped('Special types not supported by hhvm.');
```

Comment:
In fact maybe we should check [the ini setting ``hhvm.php7.scalar_types``](https://docs.hhvm.com/hhvm/configuration/INI-settings#php-7-settings) ?